tedana/decomposition/ica.py

```diff
@@ -37,16 +58,159 @@ def tedica(data, n_components, fixed_seed, maxit=500, maxrestart=10):
+            n_robust_runs=n_robust_runs,
+            max_it=maxit,
+        )
+    elif ica_method == "fastica":
+        mmix, fixed_seed = f_ica(
+            data,
+            n_components=n_components,
+            fixed_seed=fixed_seed,
+            maxit=maxit,
+            maxrestart=maxrestart,
+        )
+    else:
+        raise ValueError("The selected ICA method is invalid!")
+
+    return mmix, fixed_seed
+
+
+def r_ica(data, n_components, fixed_seed, n_robust_runs, max_it):
+    """
+    Perform robustica on `data` by running FastICA multiple times (n_robust_runs)
+    and returns mixing matrix
+
+    Parameters
+    ----------
+    data : (S x T) :obj:`numpy.ndarray`
+        Dimensionally reduced optimally combined functional data, where `S` is
+        samples and `T` is time
+    n_components : :obj:`int`
+        Number of components retained from PCA decomposition.
+    fixed_seed : :obj:`int`
+        Seed for ensuring reproducibility of ICA results.
+    n_robust_runs : :obj: `int'
+        selected number of robust runs when robustica is used. Default is 30.
+    maxit : :obj:`int`, optional
+        Maximum number of iterations for ICA. Default is 500.
+
+    Returns
+    -------
+    mmix : (T x C) :obj:`numpy.ndarray`
+        Z-scored mixing matrix for converting input data to component space,
+        where `C` is components and `T` is the same as in `data`
+    fixed_seed : :obj:`int`
+        Random seed from final decomposition.
+
+    """
+    if n_robust_runs > 200:
+        LGR.warning(
+            "The selected n_robust_runs is a very big number! The process will take a long time!"
+        )
+
+    RepLGR.info("RobustICA package was used for ICA decomposition \\citep{Anglada2022}.")
+
+    if fixed_seed == -1:
+        fixed_seed = np.random.randint(low=1, high=1000)
+
+    try:
+        rica = RobustICA(
+            n_components=n_components,
+            robust_runs=n_robust_runs,
+            whiten="arbitrary-variance",
+            max_iter=max_it,
+            random_state=fixed_seed,
+            robust_dimreduce=False,
+            fun="logcosh",
+            robust_method="DBSCAN",
+        )
+
+        s, mmix = rica.fit_transform(data)
+        q = rica.evaluate_clustering(
+            rica.S_all, rica.clustering.labels_, rica.signs_, rica.orientation_
+        )
+
+    except:
+        rica = RobustICA(
+            n_components=n_components,
+            robust_runs=n_robust_runs,
+            whiten="arbitrary-variance",
+            max_iter=max_it,
+            random_state=fixed_seed,
+            robust_dimreduce=False,
+            fun="logcosh",
+            robust_method="AgglomerativeClustering",
+        )
+
+        s, mmix = rica.fit_transform(data)
+        q = rica.evaluate_clustering(
+            rica.S_all, rica.clustering.labels_, rica.signs_, rica.orientation_
+        )
+
+    iq = np.array(np.mean(q[q["cluster_id"] >= 0].iq))  # The cluster labeled -1 is noise
```

Comment:
Personally I would have done something more exhaustive here.

For instance, if another developer were evaluating the operation of this code path, and their particular data did not yield a cluster with an ID of -1, they may erroneously conclude that something has gone wrong. It may also be unclear why, in the evaluation of the variance in ICA outcomes on a given input dataset, a cluster of decompositions explicitly flagged as noise would be explicitly excluded from a quantification of that variance.

One approach would be a more exhaustive comment, explaining how *sometimes* there is a cluster with index -1, and you have chosen to omit those data from this specific quantification for some explicit reason.

Alternatively, comments are there to explain anything that is not sufficiently clear from the code itself, so you could instead make the code more exhaustive; and at the same time, making the command output more descriptive. Two birds, one stone. You could generate a logging message stating whether the clustering of ICA results has included a "noise" cluster with index -1, how many of the total decomposition runs were allocated to that cluster, and state the fact that those outcomes will be excluded from the characterisation of index quality.

Reply:
Thanks Rob. I added a logging message to the code. However, I thought that providing the size of cluster -1 is of no actual use to the user. Moreover, I used the term outlier rather than noise which I think is more informative.